src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -27918,6 +27918,233 @@ public class C
+                Diagnostic(ErrorCode.WRN_ReturnNotNullIfNotNull, "return local1();").WithArguments("p").WithLocation(40, 13),
+                // (40,13): warning CS8825: Parameter 'pOut' must have a non-null value when exiting because parameter 'p' is non-null.
+                //             return local1(); // 1, 2
+                Diagnostic(ErrorCode.WRN_ParameterNotNullIfNotNull, "return local1();").WithArguments("pOut", "p").WithLocation(40, 13));
+        }
+
+        [Fact, WorkItem(44539, "https://github.com/dotnet/roslyn/issues/44539")]
+        public void NotNullIfNotNull_NestedLocalFunction_02()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string M0()
+    {
+        [return: NotNullIfNotNull(""p"")]
+        string? local1(string? p, [NotNullIfNotNull(""p"")] string? pOut)
+        {
+            if (p is object)
+            {
+                pOut = null;
+                return p; // 1
+            }
+
+            if (p is object)
+            {
+                pOut = p;
+                return null; // 2
+            }
+
+            if (p is object)
+            {
+                pOut = null;
+                return null; // 3, 4
+            }
+
+            if (p is null)
+            {
+                pOut = null;
+                return null;
+            }
+
+            pOut = M1();
+            return pOut; // 5, 6
+        }
+
+        return local1(""a"", null);
+    }
+
+    string? M1() => null;
+}
+";
+            var comp = CreateNullableCompilation(new[] { source, NotNullIfNotNullAttributeDefinition });
+
+            // NotNullIfNotNull enforcement doesn't work on parameters of local functions
+            // https://github.com/dotnet/roslyn/issues/47896
+            comp.VerifyDiagnostics(
+                // (19,17): warning CS8826: Return value must be non-null because parameter 'p' is non-null.
+                //                 return null; // 2
+                Diagnostic(ErrorCode.WRN_ReturnNotNullIfNotNull, "return null;").WithArguments("p").WithLocation(19, 17),
+                // (25,17): warning CS8826: Return value must be non-null because parameter 'p' is non-null.
+                //                 return null; // 3, 4
+                Diagnostic(ErrorCode.WRN_ReturnNotNullIfNotNull, "return null;").WithArguments("p").WithLocation(25, 17),
+                // (35,13): warning CS8826: Return value must be non-null because parameter 'p' is non-null.
+                //             return pOut; // 5, 6
+                Diagnostic(ErrorCode.WRN_ReturnNotNullIfNotNull, "return pOut;").WithArguments("p").WithLocation(35, 13));
+        }
+
+        [Fact, WorkItem(44539, "https://github.com/dotnet/roslyn/issues/44539")]
+        public void NotNullIfNotNull_Constructor()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public C(string? p, [NotNullIfNotNull(""p"")] string? pOut)
```

Comment:
Note: was looking for cases where a ctor chains into another ctor via `:this`.

Reply:
I added a few tests like this, but I'm not certain that it was what you were looking for. Let me know what you think.